[Tan Chee Peng] Duke Increments

t-cheepeng opened the pull request.

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

t-cheepeng committed:
Completed level 1. Greet, Echo, Exit for Duke

t-cheepeng committed:
Completed level 2. Add, List.

t-cheepeng committed:
Completed level 3. Mark as done.

t-cheepeng committed:
Completed level 4. ToDos, Events, Deadlines

t-cheepeng committed:
Completed A-TextUiTesting.

t-cheepeng committed:
Completed Level 5. Handle Errors.

t-cheepeng committed:
Completed Level 6. Delete.

t-cheepeng committed:
Completed Level 7 branch. Save.

t-cheepeng committed:
Completed Level 8 branch. Dates and Times

t-cheepeng committed:
Merge branch 'branch-Level-7'

t-cheepeng committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

t-cheepeng committed:
Completed extension A-MoreOOP and A-Packages.

t-cheepeng committed:
Merge branch 'gradle'

t-cheepeng committed:
Completed extension A-JUnit and A-Gradle.

t-cheepeng committed:
Completed A-Jar.

t-cheepeng committed:
Completed extension A-JavaDoc

t-cheepeng committed:
Completed extension A-CodingStandard.

Following Google's Java Style Guide as much as possible.

t-cheepeng committed:
Completed Level-9 Find.

t-cheepeng committed:
Merge branch 'branch-A-JavaDoc'

t-cheepeng committed:
Merge branch 'branch-A-CodingStandard'

t-cheepeng committed:
Merge branch 'branch-Level-9'.

Conflicts resolved

# Conflicts:
#	src/main/java/com/chee/io/DukePrinter.java
#	src/main/java/com/chee/parser/CommandParser.java
#	src/main/java/com/chee/parser/Verifier.java

t-cheepeng committed:
Adds basic GUI functionality to Duke.

This commit adds up to Tutorial-3 in the JavaFX tutorial for
adding a GUI to Duke. To run this portion, you are required to
change your run configurations to have it run the gradle:run task
instead of the standard java program from IntelliJ.

t-cheepeng committed:
Add GUI functionality to duke.

Previously, duke was a purely CLI program which interacted
with user through text.

With the new GUI functionality, users are now able to view a GUI
for duke which will help enhance usability and experience for them.

In the future, there is a hope to spiff up the look of the duke
GUI to make it more appealing to users.

t-cheepeng committed:
Merge branch 'branch-javafx-tutorial'

t-cheepeng committed:
Refactor code to be more readable

Previously, there were multiple checkstyle violations that were
happening in the codebase. Now, the gradle task for checkstyleMain
gives no errors and the code is conforming to the coding standard
provided by CS2103T.

t-cheepeng committed:
Adds better exception handling capabilities to Duke

With this commit, duke is now able to handle more types of events
that can cause the program to not function as expected. An additional
exception is if the stored data was somehow corrupted and its data
is not retrievable. Also, refactored out some code exception handling
code to better reflect locality of exception catching.

chowyiwai commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,59 @@
+        Command command = null;
+        try {
+            command = parser.parse(input);
+            if (command instanceof ByeCommand) {
```
Is this if condition necessary since you are going to call command.execute() either ways? You probably don't have to use instanceof too since polymorphism will ensure that the correct execute method corresponding to the type of command is called.

weigenie replied:
I agree with yiwai that this condition may not be necessary.

chowyiwai commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,59 @@
+            }
+            command.execute();
+            io.writeTasks(userHistory.getTaskList());
+        } catch (MissingInformationException | UnknownCommandException
```
I like how you have written separate Exception classes to take care of the various possible exceptions and the naming is very clear too! 👍

chowyiwai commented on src/main/java/com/chee/model/Task.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public abstract class Task {
+    String description;
+    boolean isDone;
```
I think there should be access modifiers for these.

chowyiwai commented on src/main/java/com/chee/parser/CommandParser.java:
```diff
@@ -0,0 +1,124 @@
+    public Command parse(String command)
+            throws UnknownCommandException, MissingInformationException, ParseException {
+        verifier.verify(command);
+        String[] splitWhitesapce = command.split(" ");
```
Typo in variable name?

chowyiwai reviewed (commented):
Hi Chee Peng!

I think you did a really good job so far 👍 I like how your code is separated into different classes so that each class has a very distinct purpose. Your Javadoc comments are also very elaborate but it would be good to add on and do them for all public methods. I also really like your variable names as they are very clear and it makes it easy for the reader to know what the variable represents. One thing I noticed though was that there were a few typos here and there but I think that can be easily fixed. 

I think I can learn from you in terms of separating my code into more classes so that each class is neater. Keep up the good work!

weigenie commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,58 @@
+        displayPicture.setImage(imageView);
+    }
+
+    private void flip() {
```
Perhaps you might want to consider including the javadocs for these functions.

weigenie commented on src/main/java/DialogBox.java:
```diff
@@ -0,0 +1,58 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String l, Image iv) {
```
This too.

weigenie commented on src/main/java/com/chee/commands/AddCommand.java:
```diff
@@ -0,0 +1,60 @@
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+
+public class AddCommand implements Command {
```
I like how you split all your Commands and Exceptions into smaller child classes to be more elaborate.

weigenie commented on src/main/java/com/chee/commands/AddCommand.java:
```diff
@@ -0,0 +1,60 @@
+        this.dukePrinter = dukePrinter;
+        this.dateFormat = new SimpleDateFormat("dd/MM/yyyy HHmm");
+
+        switch (commandType) {
```
Similar to how you have split the Command class into multiple child classes, perhaps you may want to split the AddCommand.

weigenie commented on src/main/java/com/chee/commands/AddCommand.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
I didn't know you could have done this!

weigenie commented on src/main/java/com/chee/commands/FindCommand.java:
```diff
@@ -0,0 +1,45 @@
+        dukePrinter.printFind(found);
+    }
+
+    private List<Task> findTasks() {
```
Could add javadocs

weigenie commented on src/main/java/com/chee/io/DukePrinter.java:
```diff
@@ -0,0 +1,158 @@
+                    + "     What can I do for you?\n"
+                    + SEPARATOR_LINE
+                    + "\n";
+    private static final String LIST_MESSAGE =
```
I like how you listed all these hardcoded messages as static final variables.

weigenie reviewed (commented):
I like the overall structure of your code and how object-oriented it is.

t-cheepeng committed:
Rename packages and refactor out AddCommand

To comply with internet standards, renamed com.chee to duke package
as domain name is not registered. Also, refactored out AddCommand
to their specific task commands for cleaner code.

t-cheepeng committed:
Add assert statements to prevent programmer mistakes. (#4)

The assert statements will ensure that the program holds up at
various points of executions. Mostly ensuring that objects being
used are non-null.

t-cheepeng committed:
Enhance readability of code across codebase

Refactor out apprpopriate portions of the codebase to add to the
readability of the codebase. Method extractions were heavily used
to comply with SLAP. Also, extra classes for
global variables were created to package code into one place.

t-cheepeng committed:
Add Travis CI to duke

A .travis.yml file is included to automate the build process
using Travis CI. The jdk version being used is jdk 11.

t-cheepeng committed:
Update travis yml file to have valid syntax

The travis yml file previously had random scripts not in a
script block.

t-cheepeng committed:
Update travis yml file for travis permission

Include a before_script to change runtime permission
for the gradlew file in the duke project so that
travis can execute the file when building.

t-cheepeng committed:
Update git gradlew permissions

t-cheepeng committed:
Corrects multiple checkstyle violations

Corrects checkstyle violations happening in the codebase.
This was alerted through the travis CI build automation used in
previous commits.

t-cheepeng committed:
Add tagging functionality to Duke

Duke is now able to handle tag commands and allows tagging of
tasks as user sees fit. Usage of the command is of the following
syntax: tag <task number> /as <tag>

Also, ToDoTest has been rewritten taking into account the tagging
functionality.

t-cheepeng committed:
Add product screenshot and user guide for product website

t-cheepeng committed:
Set theme jekyll-theme-slate

t-cheepeng committed:
Allow jar file to be cross-platform
